enhancements/sig-architecture/81-addon-lifecycle/README.md

```diff
@@ -0,0 +1,867 @@
+}
+```
+
+```golang
+// ManagedClusterAddOnSpec defines the install configuration of
+// an addon agent on managed cluster.
+type ManagedClusterAddOnSpec struct {
+  // …
+  InstallNamespace string `json:"installNamespace,omitempty"`
+	
+  // desiredVersion represents the desired addon version to install.
+  // Setting this value will trigger an upgrade or rollback (if the current version
+  // listed in the status does not match the desired version).
+  // 
+  // Valid addon versions are listed in supportedVersion in the status. 
+  // If invalid desiredVersion is specified, that may cause the upgrade or rollback 
+  // to fail.
+  // If desiredVersion is not specified, the defaultVersion in the status will be used.
+  //
+  // The progress of an update or rollback will be reported in the conditions.
+  // +optional
+  DesiredVersion string `json:"desiredVersion,omitempty"`
+
+  // …
+  Configs []AddOnConfig `json:"configs,omitempty"`
+}
+```
+
+```golang
+// ManagedClusterAddOnStatus provides information about the status of the operator.
+// +k8s:deepcopy-gen=true
+type ManagedClusterAddOnStatus struct {
+  // …
+
+  // …
+  Conditions []metav1.Condition `json:"conditions,omitempty"  patchStrategy:"merge" patchMergeKey:"type"`
+
+  // currentVersion represents the currently successfully installed addon version.
+  // If the addon installation, upgrade or rollback is in progressing, the value is unknown.
+  // +optional
+  CurrentVersion string `json:"currentVersion,omitempty"`
+
+  // lastVersion record the previous defined desiredVersion.
+  // If the previous addon desiredVersion is invalid, the value is unknown.
+  // If no previous addon desiredVersion, the value is empty.
+  // +optional
+  LastVersion string `json:"lastVersion,omitempty"`
+}
+```
+
+### New condition
+A new condition type **Progressing** will be added to the `ManagedClusterAddOn`. 
+- When the addon in installing/upgrading/rollingback, the status will be true, 
+  reason will be Installing/Upgrading/Rollingback.
+- When the addon is installed successfully/failed, the status will be false, 
+  reason will be Succeed/Failed.
+
+```yaml
+  - lastTransitionTime: "2022-11-02T05:48:12Z"
+    message: install completed with no errors.
+    reason: Succeed
+    status: "False"
+    type: Progressing
+```
+
+### Workflow
+Below is an example of `ClusterManagementAddOn`, `ManagedClusterAddOn` and `ManifestWork`:
+
+```yaml
+apiVersion: addon.open-cluster-management.io/v1alpha1
+kind: ClusterManagementAddOn
+metadata:
+  name: helloworld
+spec:
+  installStrategy:
+    type: Placements
+    placements:
+    - name: global-placement
+      namespace: default
+      addonTemplate:
+        desiredVersion: v1
+        configs:
+        - group: addon.open-cluster-management.io
+          resource: addondeploymentconfigs
+          name: global-config
+    - name: canary-placement
+      namespace: default
+      addonTemplate:
+        desiredVersion: v2
```

Comment:
As I've been playing with this to see how the CLI flow would work, I hit a snag.  What happens in the case where a managedcluster matches both global-placement and canary-placement?  The problem exists for anything setting the addonTemplate.

Also, as I've been messing with it, I've wondered what the field experience with addonDeploymentConfigs have been.  Some configuration options are definitely per-managed-cluster configuration (the enhancement explained those well), but in the field do we also see a class of "these values should be set for this addon on many clusters" and if that class exists, would we embed it here or reference it here or something else?

I'm going to push a branch that extends this PR with the kustomize example I was working with.  I don't think it's the only way, but it exposed these as issues and could drive goals, non-goals, and alternatives.

Reply:
> I'm going to push a branch that extends this PR with the kustomize example I was working with. I don't think it's the only way, but it exposed these as issues and could drive goals, non-goals, and alternatives.

I opened https://github.com/open-cluster-management-io/enhancements/pull/84 and dropped a few comment and READMEs in there to seed discussion about a possible alternative.

Reply:
we do have received many feedbacks on addonDeploymentConfigs api yet. I think currently it is mostly use as "these values should be set for this addon on many clusters". But not per cluster configuration.